File: chapter_api_tests/2024-04/validation/test_post_api-v1-projects-{projectRef}-envvars-{env}-import.py
import axios, { AxiosInstance, AxiosResponse } from 'axios';
import { describe, it, expect, beforeAll, afterAll } from '@jest/globals';

// Ensure you have the following environment variables set:
// - API_BASE_URL (e.g., https://your-api.example.com)
// - API_AUTH_TOKEN (e.g., an authentication token)

// This test suite covers the POST /api/v1/projects/:projectRef/envvars/:env/import endpoint.
// It validates request input, response structure, status codes, headers, and edge-case scenarios.

const baseURL = process.env.API_BASE_URL || 'http://localhost:3000';
const authToken = process.env.API_AUTH_TOKEN || 'YOUR_AUTH_TOKEN';

// For demonstration, we use sample values for the path parameters.
// Adjust these to match your testing environment.
const TEST_PROJECT_REF = 'sample-project';
const TEST_ENV = 'development';

// Helper method to create an Axios instance.
function createAxiosInstance(token?: string): AxiosInstance {
  return axios.create({
    baseURL,
    headers: token
      ? {
          Authorization: `Bearer ${token}`,
        }
      : {},
    validateStatus: () => true, // Let us handle status checks manually.
  });
}

// Valid payload structure assumed based on typical environment variable import:
// Adjust fields/properties as required by your actual API schema.
interface EnvVar {
  key: string;
  value: string;
}

interface ImportEnvVarsRequest {
  variables: EnvVar[];
}

// A sample valid payload.
const validPayload: ImportEnvVarsRequest = {
  variables: [
    { key: 'API_KEY', value: 'myApiKey123' },
    { key: 'SECRET', value: 'superSecretValue' },
  ],
};

// A sample invalid payload (missing "key" or "value" fields, or using the wrong types).
const invalidPayload: any = {
  variables: [
    { wrongKeyName: 'NO_KEY', wrongValueName: 123 },
  ],
};

describe('POST /api/v1/projects/:projectRef/envvars/:env/import', () => {
  let axiosInstance: AxiosInstance;

  beforeAll(() => {
    // Create an Axios instance with a valid auth token (if required by the API).
    axiosInstance = createAxiosInstance(authToken);
  });

  afterAll(() => {
    // Clean up or reset if necessary.
  });

  it('should successfully upload environment variables with a valid payload (200)', async () => {
    const url = `/api/v1/projects/${TEST_PROJECT_REF}/envvars/${TEST_ENV}/import`;

    const response: AxiosResponse = await axiosInstance.post(url, validPayload);

    // Check the expected status code.
    expect(response.status).toBe(200);

    // Check response headers.
    expect(response.headers['content-type']).toMatch(/application\/json/i);

    // Validate response body structure (assuming a success schema with a "message" or similar).
    // Replace with actual schema validations if needed.
    expect(response.data).toBeDefined();
    // For example, if #/components/schemas/SucceedResponse has a "success" field:
    // expect(response.data.success).toBe(true);
  });

  it('should return 400 or 422 for invalid payload', async () => {
    const url = `/api/v1/projects/${TEST_PROJECT_REF}/envvars/${TEST_ENV}/import`;

    const response: AxiosResponse = await axiosInstance.post(url, invalidPayload);

    // The API might return 400 or 422 for invalid requests.
    expect([400, 422]).toContain(response.status);

    // Check response headers.
    expect(response.headers['content-type']).toMatch(/application\/json/i);

    // Optional: Validate error response schema (assuming an error property exists).
    // e.g., expect(response.data).toHaveProperty('error');
  });

  it('should fail with 401 or 403 if token is missing/invalid', async () => {
    // Create an axios instance with no token (or an invalid token).
    const unauthorizedAxios = createAxiosInstance('INVALID_TOKEN');

    const url = `/api/v1/projects/${TEST_PROJECT_REF}/envvars/${TEST_ENV}/import`;

    const response: AxiosResponse = await unauthorizedAxios.post(url, validPayload);

    // The API might return 401 (Unauthorized) or 403 (Forbidden).
    expect([401, 403]).toContain(response.status);

    // Check response headers.
    // Some APIs may return text/html for error pages, so adjust accordingly.
    if (response.headers['content-type']) {
      expect(response.headers['content-type']).toMatch(/application\/json/i);
    }

    // Optionally validate error payload, if applicable.
  });

  it('should return 404 if the projectRef or env does not exist', async () => {
    // Use invalid projectRef/env to trigger 404.
    const invalidProjectRef = 'nonexistent-project';
    const invalidEnv = 'unknown-env';
    const url = `/api/v1/projects/${invalidProjectRef}/envvars/${invalidEnv}/import`;

    const response: AxiosResponse = await axiosInstance.post(url, validPayload);

    // Check that not found is returned.
    expect(response.status).toBe(404);

    // Check response headers.
    expect(response.headers['content-type']).toMatch(/application\/json/i);
  });

  it('should handle large payload without error (if supported by API)', async () => {
    // Construct a large array of environment variables.
    const largeVariables: EnvVar[] = [];
    for (let i = 0; i < 1000; i++) {
      largeVariables.push({ key: `KEY_${i}`, value: `VALUE_${i}` });
    }

    const largePayload: ImportEnvVarsRequest = {
      variables: largeVariables,
    };

    const url = `/api/v1/projects/${TEST_PROJECT_REF}/envvars/${TEST_ENV}/import`;

    const response: AxiosResponse = await axiosInstance.post(url, largePayload);

    // Depending on the API, it should return 200 if it can handle large payloads.
    // Adjust if your API returns a 413 Payload Too Large or another status.
    expect([200, 413]).toContain(response.status);

    if (response.status === 200) {
      expect(response.headers['content-type']).toMatch(/application\/json/i);
      // Validate success response...
    }
  });

  it('should handle boundary values (e.g., empty strings, special chars)', async () => {
    const boundaryPayload: ImportEnvVarsRequest = {
      variables: [
        { key: '', value: '' },          // empty
        { key: 'SPECIAL_CHARS', value: '!@#$%^&*()_+' },
      ],
    };

    const url = `/api/v1/projects/${TEST_PROJECT_REF}/envvars/${TEST_ENV}/import`;

    const response: AxiosResponse = await axiosInstance.post(url, boundaryPayload);

    // Expect success or validation error, depending on API rules.
    expect([200, 400, 422]).toContain(response.status);

    if (response.status === 200) {
      expect(response.headers['content-type']).toMatch(/application\/json/i);
      // e.g., expect(response.data.success).toBe(true);
    } else {
      expect(response.headers['content-type']).toMatch(/application\/json/i);
      // e.g., expect(response.data).toHaveProperty('error');
    }
  });

  it('should handle a server error (500) gracefully (if applicable)', async () => {
    // Triggering a 500 typically requires forcing the server to error.
    // This test is often environment-specific. We'll do a mock scenario:
    // For example, sending malicious data or something that you know triggers an error.

    const errorInducingPayload: ImportEnvVarsRequest = {
      variables: [{ key: 'FORCE_ERROR', value: 'trigger-500' }],
    };

    const url = `/api/v1/projects/${TEST_PROJECT_REF}/envvars/${TEST_ENV}/import`;

    const response: AxiosResponse = await axiosInstance.post(url, errorInducingPayload);

    // Expect 500 or another custom code if the server encountered an error.
    // If your API doesn’t actually throw 500, you can skip or adjust this test.
    expect([200, 500]).toContain(response.status);
  });
});

File: chapter_api_tests/2024-04/validation/test_post_api-v1-projects-{projectRef}-envvars-{env}-import.ts
```typescript
import axios, { AxiosInstance, AxiosResponse } from 'axios';
import { describe, it, expect, beforeAll } from '@jest/globals';

/************************************************************
 * Jest test suite for:
 * POST /api/v1/projects/{projectRef}/envvars/{env}/import
 *
 * This test suite covers:
 * 1. Input Validation
 * 2. Response Validation
 * 3. Response Headers Validation
 * 4. Edge Case & Limit Testing
 * 5. Testing Authorization & Authentication
 ************************************************************/

describe('POST /api/v1/projects/{projectRef}/envvars/{env}/import', () => {
  let client: AxiosInstance;
  let validProjectRef = 'example-project-123';
  let validEnv = 'development';

  beforeAll(() => {
    const baseURL = process.env.API_BASE_URL || 'http://localhost:3000';
    const token = process.env.API_AUTH_TOKEN || '';

    client = axios.create({
      baseURL,
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${token}`,
      },
      validateStatus: () => true, // Allow handling status codes in tests
    });
  });

  /************************************************************
   * 1) Successful upload of environment variables (200 OK)
   ************************************************************/
  it('should upload environment variables successfully (200)', async () => {
    // Example of a valid request body based on an assumed schema:
    // {
    //   vars: [
    //     { key: 'VAR_KEY', value: 'VAR_VALUE' },
    //     ...
    //   ]
    // }
    const validRequestBody = {
      vars: [
        { key: 'TEST_KEY', value: 'TEST_VALUE' },
        { key: 'ANOTHER_KEY', value: 'ANOTHER_VALUE' },
      ],
    };

    const response: AxiosResponse = await client.post(
      `/api/v1/projects/${validProjectRef}/envvars/${validEnv}/import`,
      validRequestBody
    );

    // Response Validation
    expect([200]).toContain(response.status);
    expect(response.data).toBeDefined();
    // Example: Check if response follows success structure
    // Adjust property checks based on your actual schema
    // For instance, if SucceedResponse has a "message" field:
    // expect(response.data).toHaveProperty('message');

    // Response Headers Validation
    expect(response.headers['content-type']).toMatch(/application\/json/);
  });

  /************************************************************
   * 2) Invalid request body -> expect 400 or 422
   ************************************************************/
  it('should return 400 or 422 for invalid request body', async () => {
    const invalidRequestBody = {
      // Missing or malformed "vars" field
      // e.g., string instead of array
      vars: "this-should-be-an-array",
    };

    const response: AxiosResponse = await client.post(
      `/api/v1/projects/${validProjectRef}/envvars/${validEnv}/import`,
      invalidRequestBody
    );

    // Expecting 400 or 422 for invalid payload
    expect([400, 422]).toContain(response.status);
    expect(response.data).toBeDefined();

    // Check if error response structure matches expectations (e.g., error details)
    // Example:
    // expect(response.data).toHaveProperty('error');

    // Response Headers Validation
    expect(response.headers['content-type']).toMatch(/application\/json/);
  });

  /************************************************************
   * 3) Unauthorized or forbidden -> expect 401 or 403
   ************************************************************/
  it('should return 401 or 403 when the Authorization token is missing or invalid', async () => {
    // Create a client without auth token
    const unauthorizedClient = axios.create({
      baseURL: process.env.API_BASE_URL || 'http://localhost:3000',
      headers: {
        'Content-Type': 'application/json',
      },
      validateStatus: () => true,
    });

    const validRequestBody = {
      vars: [{ key: 'KEY_NO_AUTH', value: 'VALUE_NO_AUTH' }],
    };

    const response: AxiosResponse = await unauthorizedClient.post(
      `/api/v1/projects/${validProjectRef}/envvars/${validEnv}/import`,
      validRequestBody
    );

    // Expecting 401 or 403
    expect([401, 403]).toContain(response.status);
    // Check if response content matches expected structure
    // For example:
    // expect(response.data).toMatchObject({ error: expect.any(String) });
  });

  /************************************************************
   * 4) Resource not found -> expect 404
   ************************************************************/
  it('should return 404 when projectRef or env does not exist', async () => {
    const invalidProjectRef = 'non-existing-project';
    const invalidEnv = 'non-existing-env';

    const validRequestBody = {
      vars: [{ key: 'TEST_KEY_404', value: 'TEST_VALUE_404' }],
    };

    const response: AxiosResponse = await client.post(
      `/api/v1/projects/${invalidProjectRef}/envvars/${invalidEnv}/import`,
      validRequestBody
    );

    // Expecting 404
    expect(response.status).toBe(404);
    // Check response structure if applicable
    // Example:
    // expect(response.data).toHaveProperty('error');

    // Response Headers Validation
    expect(response.headers['content-type']).toMatch(/application\/json/);
  });

  /************************************************************
   * 5) Edge case: Empty request body
   *    - Depending on API constraints, expect 400, 422, or success if empty is allowed
   ************************************************************/
  it('should handle empty request body', async () => {
    const emptyRequestBody = {};

    const response: AxiosResponse = await client.post(
      `/api/v1/projects/${validProjectRef}/envvars/${validEnv}/import`,
      emptyRequestBody
    );

    // Expecting 400 or 422 if empty requests are invalid
    // or possibly 200 if the API allows an empty import
    expect([200, 400, 422]).toContain(response.status);

    // If success is valid for empty imports, we can check success structure;
    // otherwise, check error.
    // expect(response.data).toHaveProperty('error'); or similar.

    expect(response.headers['content-type']).toMatch(/application\/json/);
  });

  /************************************************************
   * 6) Edge case: Large payload
   *    - Test the API handling of large imports.
   ************************************************************/
  it('should handle a large payload of environment variables', async () => {
    // Creating a large array of environment variables
    const largeVars = Array.from({ length: 1000 }, (_, index) => ({
      key: `LARGE_KEY_${index}`,
      value: `LARGE_VALUE_${index}`,
    }));

    const largeRequestBody = {
      vars: largeVars,
    };

    const response: AxiosResponse = await client.post(
      `/api/v1/projects/${validProjectRef}/envvars/${validEnv}/import`,
      largeRequestBody
    );

    // Expect success or appropriate handling (e.g., 413 if the payload is too large)
    expect([200, 400, 413, 422]).toContain(response.status);

    expect(response.headers['content-type']).toMatch(/application\/json/);
  });

  /************************************************************
   * 7) Malformed request (simulate server error handling)
   *    - If relevant, we can test for 500 or other 5xx.
   ************************************************************/
  it('should handle server errors (simulate a malformed request that leads to 500)', async () => {
    // This test depends on whether the server can produce a 500
    // For demonstration, we pass an obviously incorrect structure.

    const malformedBody = {
      vars: 12345, // Not an array or object structure as expected
    };

    const response: AxiosResponse = await client.post(
      `/api/v1/projects/${validProjectRef}/envvars/${validEnv}/import`,
      malformedBody
    );

    // Some servers may return 400 or 422 instead of 500 for malformed bodies.
    // If your server can return 500, adjust the test accordingly.
    expect([400, 422, 500]).toContain(response.status);

    expect(response.headers['content-type']).toMatch(/application\/json/);
  });
});
```
